### PR TITLE
Provide Model entity data.

### PR DESCRIPTION
**Pull request by julien-me**
This PR is concerned with Model entities only.

- [x] Read the received entity data.
- [x] Notify and pass the entity data to the user application.
- [x] Update `ENTITY_SERVER.md`.
- [x] List entities (Id, type, position) in the example app.

### CODE REVIEW

File: src/domain/networking/packets/EntityData.ts [hunk 1]
```diff
@@ -9,29 +9,1932 @@
...
+
+                    const entityTypeCodec = new ByteCountCoded();
+                    let encodedData = new DataView(entityData.buffer, entityData.byteOffset + entityDataPosition);
+                    entityDataPosition += entityTypeCodec.decode(encodedData, encodedData.byteLength);
+
+                    const entityType = entityTypeCodec.data;
+
+                    if (entityType !== EntityTypes.Model) {
+                        console.error("Entity is not a Model entity");
+                        return [];
```
Comment (ctrlaltdavid) [inline src/domain/networking/packets/EntityData.ts:921]: This is going to return from the read() method immediately without any entity data, even if Model entities have been read. Instead, it should skip over the current entity's data and continue on with reading the next.

File: src/domain/octree/OctreePacketProcessor.ts [hunk 2]
```diff
@@ -70,30 +72,50 @@ class OctreePacketProcessor {
...
-            // WEBRTC TODO: Address further C++ code.
-            return;
+        let messageLocal = message;
+
+        const octreePacketType = messageLocal.getType();
+
+        if (octreePacketType === PacketType.OctreeStats) {
+            // WEBRTC TODO: Address further C++ code - process OctreeStats packet.
+
+            console.error("OctreePacketProcessor: Packet type not processed: ", octreePacketType);
```
Comment (ctrlaltdavid) [inline src/domain/octree/OctreePacketProcessor.ts:92]:
```suggestion
```
Let's remove this console error. I don't think the OctreeStats packet is particularly important to address in the near future.

File: src/domain/octree/OctreePacketProcessor.ts [hunk 3]
```diff
@@ -70,30 +72,50 @@ class OctreePacketProcessor {
...
         switch (packetType) {
-            case PacketType.EntityData:
-                PacketScribe.EntityData.read();
-                // WEBRTC TODO: Emit data.
-                this.#_addedEntity.emit();
+            case PacketType.EntityData: {
+                const entityDataDetails = PacketScribe.EntityData.read(messageLocal.getMessage());
+                this.#_entityData.emit(entityDataDetails);
                 break;
+            }
```
Comment (ctrlaltdavid) [inline src/domain/octree/OctreePacketProcessor.ts:118]: Let's make the console.error() report just EntityErase condition, and explicitly say "... not processed: EntityErase".

Also, a uint16_t needs to be read / skipped over for the EntityQueryInitialResultsComplete case.

File: package.json [hunk 4]
```diff
@@ -61,8 +57,10 @@
         "wrtc": "^0.4.7"
     },
     "dependencies": {
+        "@types/pako": "^2.0.0",
```
Comment (ctrlaltdavid) [inline package.json:60]: This probably just needs to be in devDependencies.

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]

File: src/domain/networking/packets/EntityData.ts (hunk 1 above)
Comment (julien-me) [reply src/domain/networking/packets/EntityData.ts:921]: The issue then is, as far as I can tell, there's no way to know how many bytes to skip.

There's an `EntityItem::expectedBytes()` (see link below) but it doesn't help. It's actually more of a `minimumExpectedBytes()` than a true count of the expected bytes.
https://github.com/vircadia/vircadia/blob/master/libraries/entities/src/EntityTree.cpp#L197

Nor does it appear to be encoded anywhere in the data.

I could break from the loop maybe. And return what has already been extracted. It would cut at the first non Model entity.

File: src/domain/networking/packets/EntityData.ts [hunk 5]
```diff
@@ -9,29 +9,1932 @@
...
+                            animationPlaying,
+                            animationLoop,
+                            animationFirstFrame,
+                            animationLastFrame,
+                            animationHold
+                        });
+                }
+                dataPosition += sectionLength;
+            }
+            dataPosition += entityMessagePosition;
```
Comment (ctrlaltdavid) [inline src/domain/networking/packets/EntityData.ts:1930]: Having both of these updates to dataPosition isn't right?

Also, should assert that `sectionLength` matches up with entityMessagePosition, to help catch errors when implementing other entity types.

File: src/domain/networking/packets/EntityData.ts (hunk 1 above)
Comment (ctrlaltdavid) [reply src/domain/networking/packets/EntityData.ts:921]: Possible code:
```
                    if (entityType !== EntityTypes.Model) {
                        dataPosition += 2;  // sectionLength bytes
                        dataPosition += sectionLength;
                        continue;
                    }

```

File: src/domain/networking/packets/EntityData.ts [hunk 6]
```diff
@@ -9,31 +9,2064 @@
...
+                useOriginalPivot,
+                animationUrl,
+                animationAllowTranslation,
+                animationFPS,
+                animationFrameIndex,
+                animationPlaying,
+                animationLoop,
+                animationFirstFrame,
+                animationLastFrame,
+                animationHold
```
Comment (ctrlaltdavid) [inline src/domain/networking/packets/EntityData.ts:2024]: Animation properties should be grouped in a `animation` property, like the native scripting API is: https://apidocs.vircadia.dev/Entities.html#.EntityProperties-Model

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]

File: src/domain/shared/ByteCountCoding.ts [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+// WEBRTC TODO: Make the class generic.
+class ByteCountCoded {
+    // C++ template<typename T> class ByteCountCoded
+
+    // WEBRTC TODO: Move to NumericalConstants.ts
+    readonly #BITS_IN_BYTE = 8;
+
+    #_data = 0;
+
+    get data(): number {
```
Comment (ctrlaltdavid) [inline src/domain/shared/ByteCountCoding.ts:27]: This property needs JSDoc.

File: src/domain/shared/ByteCountCoding.ts [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+        return this.#_data;
+    }
+
+    /*@devdoc
+     *  Decode the encoded data.
```
Comment (ctrlaltdavid) [inline src/domain/shared/ByteCountCoding.ts:32]:
```suggestion
     *  Decodes the byte count encoded data.
```

File: src/domain/shared/ByteCountCoding.ts [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+//  Copyright 2022 Vircadia contributors.
+//  Copyright 2022 DigiSomni LLC.
+//
+//  Distributed under the Apache License, Version 2.0.
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+
+
+/*@devdoc
+ *  The <code>ByteCountCoded</code> class provides facilities to decode data.
```
Comment (ctrlaltdavid) [inline src/domain/shared/ByteCountCoding.ts:14]:
```suggestion
 *  The <code>ByteCountCoded</code> class provides facilities to decode byte count encoded data.
```

File: src/domain/networking/NLPacket.ts [hunk 10]
```diff
@@ -180,6 +196,19 @@ class NLPacket extends Packet {
             this._messageData.dataPosition = headerStart;  // Reset to start of header for #localHeaderSize().
             this.adjustPayloadStartAndCapacity(NLPacket.#localHeaderSize(this._messageData.type));
 
+        } else if (param0 instanceof DataView && typeof param1 === "number" && param2 instanceof SockAddr) {
```
Comment (ctrlaltdavid) [inline src/domain/networking/NLPacket.ts:199]: The constructor documentation needs updating with this variation. The following JSDoc does this:
```
 *  @class NLPacket
 *  @extends Packet
 *  @param {PacketType|DataView|Packet} type|packet - The type of NLPacket to create.
 *      <p>A Dataview to create the NLPacket from.</p>
 *      <p>A base Packet to create the NLPacket from.</p>
 *      <p>Note: The {@link MessageData} from the base packet is reused in-place, not copied.</p>
 *  @param {number|number|unused} size - The size of the packet in bytes. If <code>-1</code>, a packet of the maximum size is
 *      created (though not all of it need be sent). <strong>Default Value:</strong> <code>-1</code>
 *      <p>The size of the packet in bytes.
 *      <p>Unused.</p>
 *  @param {boolean|SockAddr|unused} isReliable - <code>true</code> if the packet is to be sent reliably, <code>false</code> if
 *      it isn't. <strong>Default Value:</strong> <code>false</code
 *      <p>The sender's IP address and port.</p>
 *      <p>Unused.</p>
 *  @param {boolean|unused|unused} isPartOfMessage - <code>true</code> if the packet is part of a message, <code>false</code> if
 *      it isn't. <strong>Default Value:</strong> <code>false</code>
 *      <p>Unused.</p>
 *      <p>Unused.</p>
 *  @param {PacketVersion|unused|unused} version - The version of the NLPacket to create. <strong>Default Value:</strong>
 *      <code>0</code>
 *      <p>Unused.</p>
 *      <p>Unused.</p>
```

File: src/domain/networking/packets/EntityData.ts [hunk 11]
```diff
@@ -9,31 +9,2064 @@
...
+        // WEBRTC TODO: Read and use the variable colorInPacketMask.
+        dataPosition += 1;
+
+        // WEBRTC TODO: Do not hardcode value.
+        // eslint-disable-next-line
+                // @ts-ignore
+        const bytesForMask = 2;
+        dataPosition += bytesForMask;
+
+        // WEBRTC TODO: Address further C++ code.
```
Comment (ctrlaltdavid) [inline src/domain/networking/packets/EntityData.ts:921]: I think that this code may come into play when either (a) there are more entities in a domain than just one or two and they're store in octree nodes, or (b) entities are parented to others. In either case, there is some recursive code in the C++ that reads the entity data which we're probably going to need to implement ASAP.
This can perhaps be handled as a separate task (https://github.com/vircadia/vircadia-web-sdk/issues/131) - e.g., in parallel to reading the second entity type (https://github.com/vircadia/vircadia-web-sdk/issues/130).

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]

File: src/domain/networking/packets/EntityData.ts [hunk 12]
```diff
@@ -9,31 +9,2064 @@
...
+        // 1 represents sizeof(unsigned char) in the C++ code.
+        if (data.byteLength - dataPosition < 1) {
+            console.error("Not enough meaningful data.");
+            return {
+                bytesRead: dataPosition - position,
+                entitiesDataDetails: []
+            };
+        }
+
+        // WEBRTC TODO: Read and use the variable colorInPacketMask.
```
[the remark this answers is not in the document]
Comment (julien-me) [reply src/domain/networking/packets/EntityData.ts:912]: Seems correct.

File: src/domain/octree/OctreePacketProcessor.ts [hunk 13]
```diff
@@ -75,33 +77,53 @@ class OctreePacketProcessor {
...
                 console.warn("OctreePacketProcessor: OctreeStats packet not processed.");
                 this.#_haveWarnedOctreeStats = true;
             }
-            // WEBRTC TODO: Address further C++ code.
-            return;
+
+            // WEBRTC TODO: Address further C++ code - process OctreeStats packet.
+
+            // WEBRTC TODO: Do not hardcode statsMessageLength.
+            const statsMessageLength = 222;
```
[the remark this answers is not in the document]
Comment (julien-me) [reply src/domain/octree/OctreePacketProcessor.ts:103]: Yes. It always reads the same 222 bytes.
It happens in this function:  https://github.com/vircadia/vircadia/blob/master/libraries/octree/src/OctreeSceneStats.cpp#L342-L399

File: src/domain/networking/packets/EntityData.ts [hunk 14]
```diff
@@ -9,31 +9,2064 @@
...
+            };
+        }
+
+        const entitiesDataDetails: EntityDataDetails[] = [];
+
+        for (let i = 0; i < numberOfEntities; i++) {
+            const entityItemID = new Uuid(data.getBigUint128(dataPosition, UDT.BIG_ENDIAN));
+            dataPosition += 16;
+
+            const entityTypeCodec = new ByteCountCoded();
```
[the remark this answers is not in the document]
Comment (julien-me) [reply src/domain/networking/packets/EntityData.ts:968]: Yes. Fixed.

File: src/domain/ENTITY_SERVER.md [hunk 15]
```diff
@@ -91,11 +95,9 @@ Sent by `EntityServer.#sendNackPackets()` which is called in `EntityServer.updat
 
 This is handled by `OctreePacketProcessor.handleOctreePacket()`. The packet is not queued for later processing; instead, it is immediately processed by `OctreePacketProcessor.processPacket()`.
 
-The user application is notified of the new or changed entity via the `EntityServer.entityData` signal. This signal (unlike the scripting API's `Entities.addingEntity` signal) includes all the entity data (not just the entity ID) because the user application will want all the data in order to render it.
-
-_TO DO: Instead of an `EntityServer.entityData` signal should there be `entityAdded` and `entityUpdated` signals?_
+The user application is notified of the new or changed entity via the `EntityServer.entityData` signal. This signal (unlike the scripting API's `Entities.addingEntity` signal) includes all the entity data (not just the entity ID) because the user application will want all the data in order to render it. When an entity is edited, all data concerning this entity is passed to the user application.
```
[the remark this answers is not in the document]
Comment (julien-me) [reply src/domain/ENTITY_SERVER.md:98]: It sends all the properties.

File: tests/domain/shared/ByteCountCoding.unit.test.js [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+//
+
+import ByteCountCoded from "../../../src/domain/shared/ByteCountCoding";
+
+
+describe("ByteCountCoding - unit tests", () => {
+
+    /* eslint-disable @typescript-eslint/no-magic-numbers */
+
+    test("Can decode an encoded byte", () => {
```
[the remark this answers is not in the document]
Comment (julien-me) [reply tests/domain/shared/ByteCountCoding.unit.test.js:19]: I haven't identified edge cases. Maybe I would need to understand `encode`.

File: src/domain/networking/packets/EntityData.ts [hunk 17]
```diff
@@ -9,31 +9,2063 @@
...
+    jointRotations: quat[] | undefined;
+    jointTranslationsSet: boolean[] | undefined;
+    jointTranslations: vec3[] | undefined;
+    groupCulled: boolean | undefined;
+    relayParentJoints: boolean | undefined;
+    blendShapeCoefficients: string | undefined;
+    useOriginalPivot: boolean | undefined;
+} & AnimationProperties;
+
+type EntityDataDetails = CommonEntityProperties & ModelEntityProperties;
```
Comment (ctrlaltdavid) [inline src/domain/networking/packets/EntityData.ts:557]:
```suggestion
    blendShapeCoefficients: string | undefined;
    useOriginalPivot: boolean | undefined;
    animation: AnimationProperties | undefined;
};

type EntityDataDetails = CommonEntityProperties & ModelEntityProperties;
```
I was thinking have an `animation` property which is an object. This makes it the same as the "animation" property in the native scripting interface: https://apidocs.vircadia.dev/Entities.html#.EntityProperties-Model

File: src/domain/octree/OctreePacketProcessor.ts [hunk 18]
```diff
@@ -72,36 +74,57 @@ class OctreePacketProcessor {
...
             }
-            // WEBRTC TODO: Address further C++ code.
-            return;
+
+            // WEBRTC TODO: Address further C++ code - process OctreeStats packet.
+
+            // The stats message is always 222 bytes long.
+            const statsMessageLength = 222;
+            // eslint-disable-next-line @typescript-eslint/no-magic-numbers
+            const piggybackBytes = messageLocal.getMessage().byteLength - statsMessageLength;
```
Comment (ctrlaltdavid) [inline src/domain/octree/OctreePacketProcessor.ts:106]:
```suggestion
            const STATS_MESSAGE_LENGTH = 222;
            const piggybackBytes = messageLocal.getMessage().byteLength - STATS_MESSAGE_LENGTH;
```
Constants upper-case per C++ coding standard.

File: src/EntityServer.ts [hunk 19]
```diff
@@ -126,11 +125,12 @@ class EntityServer extends AssignmentClient {
     }
 
     /*@sdkdoc
-     *  Triggered when an entity data packet is received.
-     *  @callback EntityServer~AddedEntityCallback
+     *  Triggered when new or changed entity data is received from the entity server.
+     *  @callback EntityServer~entityData
+     *  @param {EntityDataDetails[]} entityData - The entity data information for one or more entities.
```
Comment (ctrlaltdavid) [inline src/EntityServer.ts:130]:
```suggestion
     *  @param {EntityDataDetails[]} entityData - The entity data for one or more entities. Note that complete entity data is provided for both new and changed entities.
```

File: src/domain/networking/packets/EntityData.ts [hunk 20]
```diff
@@ -9,31 +9,2063 @@
...
+     *     played are maintained when the animation stops playing, <code>false</code> if they aren't.
      */
 
     /*@devdoc
-     *  Reads an {@link PacketType(1)|EntityData} packet.
+     *  Reads an {@link PacketType(1)|EntityData} packet containing the details of one or more entities.
      *  @function PacketScribe.EntityData&period;read
-     *  @returns {PacketScribe.EntityDataDetails} The entity data information.
+     *  @read {DataView} data - The {@link Packets|EntityData} message data to read.
+     *  @returns {PacketScribe.EntityDataDetails[]} The entity data information for one or more entities.
```
Comment (ctrlaltdavid) [inline src/domain/networking/packets/EntityData.ts:791]:
```suggestion
     *  @returns {PacketScribe.EntityDataDetails[]} The entity data for one or more entities.
```

File: src/domain/octree/OctreePacketProcessor.ts [hunk 21]
```diff
@@ -72,36 +74,57 @@ class OctreePacketProcessor {
 
     /*@devdoc
      *  Triggered when an entity data packet is received.
-     *  @function OctreePacketProcessor.addedEntity
+     *  @callback OctreePacketProcessor.entityData
+     *  @param {EntityDataDetails[]} entityData - The entity data information for one or more entities.
```
Comment (ctrlaltdavid) [inline src/domain/octree/OctreePacketProcessor.ts:78]:
```suggestion
     *  @param {EntityDataDetails[]} entityData - The entity data for one or more entities.
```

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]
